File: P2L1.m
function [Mxy, Mz] = P2L1(vars,fdv)
%Format: function [Mxyev, Mxyiv, vb, Mev, Miv] = P2LBv4(vars,fdv)
%
%input:
    % vars:                 values for parms have been or will be fit
    % fdv = "forward variables" containing info about the measurement
    %    fdv.fitvars        Names of parms that need to be fit
    %    fdv.knowns         Names of parms that are known
    %    fdv.knownvals      Values of known parameters
    %    fdv.FlipAngle      Excitation angle list [2 x NFlips] in degrees
    %    fdv.TR:            Repetition time [2 x NFlips]
    %    fdv.taxis:         time axis for the data [1 x NFlips]
    %    fdv.data           [2 x NTP] observed [pyr;lac] data
    %    fdv.VIFP           *** NOT USED FOR THIS MODEL ***
    %    fdv.verbose        =1 to plot results of this script
 
%output:
    % Mxy                   [2 x NTP] Mxy observed [Pyr;Lac] 
    % Mz                    [2 x NTP] Mz [Pyr;Lac]
%
% Note Mz reflects magnetization before excitation pulse,  
%  while Mxy reflects transverse magnetization due to pulse.
%
% Note total observed signal is [Pyr;Lac]
%

    eps=1e-6;
       
    %Unpack variables:
    for ii=1:length(vars)
        x.(fdv.fitvars{ii})=vars(ii);
    end
    for ii=1:length(fdv.knowns)
        x.(fdv.knowns{ii})=fdv.knownvals(ii);
    end

    %Data from each excitation:
    PxySeg  = fdv.data(1,:);
    PzSeg   = PxySeg./sind(fdv.FlipAngle(1,:));
    
    %Initialize return variables:
    Mxy  = zeros(2,fdv.ntp);
    Mz   = zeros(2,fdv.ntp);

    %Set up calculation
    % Lac = Lac0*exp(At) + kpl*integral(0,t:exp(A*(t-tau))*[Pyr(tau)]*dtau)
    % Here, A is a single value:
    A = -(x.klp + (1/x.T1Lac));
    
    %Calculate signal evolution each TR:
    
    LzSegIC = x.L0; %Set Initial Conditions
    for ii=1:fdv.ntp
        LxySeg=zeros(1,fdv.NSeg);
        LzSeg =zeros(1,fdv.NSeg);
        for jj=1:fdv.NSeg
            iiseg=(ii-1)*(fdv.NSeg)+jj; % Works when NSeg=1 too
            TR=fdv.TR(iiseg);
            %
            %First account for EV signal already present and its evolution
            %
            %Longitudinal magnetization available at the start of each segment:
            LzSeg(jj)=LzSegIC;
            %Signal observed at each excitation, at start of segment:
            LxySeg(jj)=LzSegIC.*sind(fdv.FlipAngle(2,iiseg));
            %At the end of this TR, Mz evolves to contribute to IC for next:
            if iiseg<fdv.NFlips % Don't need to calc after last datapoint
                LzSeg1=exp(A*TR)*LzSegIC*cosd(fdv.FlipAngle(2,iiseg));
                %MevSeg1=(exp(dD*TR)).*(P\(MevSegIC.*cosd(fdv.FlipAngle(:,iiseg))));
                %
                %Now calculate new spins flowing into the system during this TR:
                %
                %Get slope and y-intercept for pyruvate forcing function:
                b=PzSeg(1,iiseg);
                m=(PzSeg(1,iiseg+1)-PzSeg(1,iiseg))/TR;
                %At the end of this TR, inflowing spins will lead to:
                LzSeg2= (((m/A+b)*exp(A*TR)) - (m*(TR+1/A)) - b)/A;
                %Total signal at end of TR equals IC for next TR:
                LzSegIC = LzSeg1 + (x.kpl)*LzSeg2;
            end
        end
        if (fdv.NSeg>1)
            %Note that this calculates the average signal from a segmented
            %acquisition.  May need to modify or add weighting factors to 
            %compensate for k-space coverage.
            %
            %Pyr
            Mxy(1,ii)=mean(PxySeg(1,(ii-1)*fdv.NSeg+1:ii*fdv.NSeg),2);
            Mz(1,ii)=mean(PzSeg(1,(ii-1)*fdv.NSeg+1:ii*fdv.NSeg),2);
            %Lac
            Mxy(2,ii)=mean(LxySeg);
            Mz(2,ii)=mean(LzSeg);
        else
            Mxy(:,ii)=[PxySeg(1,ii); LxySeg];
            Mz(:,ii)=[PzSeg(1,ii) ; LzSeg];
        end            
    end

if fdv.verbose
    figure(99)
    plot(fdv.taxis(1:fdv.NSeg:end),Mxy(1,:),'g-',...
        fdv.taxis(1:fdv.NSeg:end),Mxy(2,:),'b-')
    legend('Pyr','Lac')
    xlabel('Time (s)')
    grid on
end    

File: python/P2L1.py
import numpy as np
import matplotlib.pyplot as plt

def P2L1(vars, fdv):

    '''
    input:
        vars:               Value for parameter to be fit as array of size (1,)
        fdv:                Dictionary with the following keys
            fdv['fitvarNames']      Names of variables to be fit as string
            fdv['knowns']           Dictionary object containing known values
                                        (see TestP2L1.py for default)
            fdv['ntp']              Number of time points as int
            fdv['NSeg']             Number of segments per time point as int
                                        Note: for version 0.0.1, NSeg = 1
            fdv['NFlips']           Total number of excitations = ntp * NSeg as int
            fdv['FlipAngle']        Numpy array of excitation angles in degrees size = (2, NFlips)
            fdv['TR']               Single valued numpy array of repetition time size = (1, NFlips)
            fdv['data']             Observed data in the format [ [pyr], [lac] ] with size = (2, ntp)
            fdv['taxis']            Time axis used for the data with size = (1, NFlips)
            fdv['Name']             Title for the output. Only affects plot title on TestP2L1.py
            fdv['verbose']          Bool type. True plots results of P2L1 directly

    output:
        Mxy:                Mxy observed [ [Pyr], [Lac] ] with shape = (2, ntp)
        Mz:                 Mz [ [Pyr], [Lac] ] with shape = (2, ntp)

    Note: Mz is the magnetization before the excitation pulse,
        while Mxy is the transverse magnetization due to the pulse

    Total observed signal is [ [Mxy], [Mz] ]
    '''

    # Unpack fit variables into better naming scheme
    i = 0
    for item in fdv['fitvarNames']:
        fdv[item] = vars[i]
        i += 1
        
    # Data from each excitation
    PxySeg = fdv['data'][0]
    PzSeg = np.divide( PxySeg, np.sin( np.radians(fdv['FlipAngle'][0]) ))
    
    # Initialize return variables 
    Mxy = np.zeros( (2, fdv['ntp']) )
    Mz = np.zeros( (2, fdv['ntp']) )

    # Set up for the following equation
    #  Lac = Lac0 * exp(At) + kpl * inte(0,t: exp(A*(t-T)) *[Pyr(T)] *dT)
    A = -(fdv['knowns']['klp'] + (1/fdv['knowns']['T1Lac']))

    #### Caluclate signal evolution in each TR ###
    LzSegIC = fdv['knowns']['L0']  # Initial condition before acquisition starts
    for i in range(0, fdv['ntp']):
        LxySeg = np.zeros( (1, fdv['NSeg']) )
        LzSeg = np.zeros( (1, fdv['NSeg']) )
    
        for j in range(0, fdv['NSeg']):
            iSeg = (i)*(fdv['NSeg'])+j
            TR = fdv['TR'][0][iSeg]
    
            # First account for signal already in the slice and its evolution

            # Longitudinal M available at start of each cycle
            LzSeg[j] = LzSegIC
            # Signal observed at start of each cycle
            LxySeg[j] = np.multiply(LzSegIC, np.sin( np.radians(fdv['FlipAngle'][1, iSeg]) ))

            # Evolution of this cycle becomes the IC of the next
            if iSeg < fdv['NFlips']: # No evolution after last datapoint
                LzSeg1 = np.exp(A*TR)*LzSegIC*np.cos(
                    np.radians(fdv['FlipAngle'][1, iSeg] ))

                # Now account for new spins flowing nto the system during TR

                # Obtain parameters for linear pyruvate forcing function
                b = PzSeg[iSeg-1]
                m = (PzSeg[iSeg]-PzSeg[iSeg-1]) / TR 

                # Contribution from inflowing spins during this TR
                LzSeg2 = (((m/A + b)* np.exp(A*TR)) - (m*(TR+1/A)) - b)/A

                #Total signal at the end of TR is sum of inflowing and already present
                LzSegIC = LzSeg1 + fdv['kpl'] * LzSeg2
                
        
        if fdv['NSeg']>1:
            # In version 0.0.1, fdv['NSeg']>1 will never be evaluated
            print('Segmented acquisition not available in version 0.0.1')
            #Pyr
            Mxy[0, i-1] = np.mean(
                PxySeg[range(fdv['NSeg']*(i-1)+ 1, i*fdv['NSeg'])], 2)
            Mz[0, i-1] = np.mean( 
                PzSeg[range(fdv['NSeg']*(i-1)+ 1, i*fdv['NSeg'])], 2)

            #Lac
            Mxy[1:i-1] = np.mean(LxySeg)
            Mz[1:i-1] = np.mean(LzSeg)
        else:
            Mxy[:,i-1] = [PxySeg[i-1], LxySeg[0][0]]
            Mz[:,i-1] = [PzSeg[i-1], LzSeg[0][0]]
    ### END OF CALCULATION LOOP ###

    if fdv['verbose']:
        fig, ax = plt.subplots(figsize = (12, 6))

        ax.plot(fdv['taxis'], Mxy[0], color = 'green', label = 'Pyr')
        ax.plot(fdv['taxis'], Mxy[1], color = 'blue', label = 'Lac')

        ax.grid()
        ax.legend(fontsize = 'x-large')

        ax.set_xlabel('Time (s)', fontsize = 'x-large')

        plt.show()

    return (Mxy, Mz)